# CMSC478 Machine Learning

# Assignment-1: Binary Classification and Linear Regression

*Connor Ragan AN72374* 

## Overview and Learning Objectives

This is your first ML assignment! So far, you've learned about the basic and fundamental concepts in Machine Learning such as supervised learning (e.g. classification) vs unsupervised learning (clustering), regression, overfitting, and model evaluation. You learned that one of the most common approaches in supervised learning is classification and regression. In this assignment, you are going to build and train a classification model for diagnosing breast cancer, as well as a regression model and how it is impacted by overfitting.

Pedagogically, this assignment will help you:
- better understand the concepts you learned and how to use ML models in pratice. 
- brush up your Python skills - and possibly learn a couple of new "Pythonic" tricks!
- pratice reading documentation. This is a very important skill in AI/ML/Data Science collaborative environments and teams.

So, let's get started!

## Part I - Binary Classification

Binary classification means you are classifying only two classes/labels. You are going to build a binary classifier that can classify breats cancer

The first thing to do is importing the breast cancer data set from scikit learn. It is one of the datasets that scikit learn provides for practicing machine learning.

In [1]:
from sklearn.datasets import load_breast_cancer

In [2]:
# Load dataset by calling the related function from what you just imported
data = load_breast_cancer()

Now, let's see what is the type of the variable `data` which holds our data.

In [3]:
type(data)

sklearn.utils.Bunch

`Bunch` is a container object for datasets. It works like a [dictionary data structure in Python](https://docs.python.org/3/tutorial/datastructures.html#dictionaries). Since this is a classification (supervised learning) ML model, you have bothe the data and the labels for training and testing, and everything has been stored in `data` variable. We can extract features and labels from it.

In [4]:
# Organize our data
label_names = data['target_names']
labels = data['target']
feature_names = data['feature_names']
features = data['data']

In [5]:
print(label_names)
print(labels[0])
print(feature_names[0])
print(features[0])

['malignant' 'benign']
0
mean radius
[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
 1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
 6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
 1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
 4.601e-01 1.189e-01]


The two classes (aka labels) here are `malignant` and `benign` which refers to the tumors you are going to classify. They are represented in the labels as 0 and 1 respectively.

As mentioned in the lectures, the data for supervised learning (both classification and regression) is split into training set and test set. We do this by importing another module from scikit learn.

In [6]:
from sklearn.model_selection import train_test_split

And now, we can split the data. The split ratio we are going to choose is 0.7 for training and 0.3 for testing, but we only need to specify one of them `test_size` here.

In [7]:
train, test, train_labels, test_labels = train_test_split(features, labels, test_size=0.30, random_state=42)

And, we can now choose which classifier from the built-in classifiers in sklearn we want to choose. We are going to choose Stochastic Gradient Descent Classifier (SGDC) similar to what we did in the lecture.

In [8]:
from sklearn.linear_model import SGDClassifier

Now complete the following cell checking how you should do it from the textbook.

## <font color="red"> Required Coding

In [37]:
### START CODING HERE ###

# Call SGDClassifier with a random_state of 42
sgd_clf = SGDClassifier(random_state=42000)

# Fit the model to train and train_labels
sgd_clf.fit(train, train_labels)

### END CODING HERE ###

SGDClassifier(alpha=0.0001, average=False, class_weight=None,
              early_stopping=False, epsilon=0.1, eta0=0.0, fit_intercept=True,
              l1_ratio=0.15, learning_rate='optimal', loss='hinge',
              max_iter=1000, n_iter_no_change=5, n_jobs=None, penalty='l2',
              power_t=0.5, random_state=42000, shuffle=True, tol=0.001,
              validation_fraction=0.1, verbose=0, warm_start=False)

## <font color="red"> Required Coding

In [38]:
# Now, make predictions with the model you just trained on test and store the results in test_predictions

### START CODING HERE ###
test_predictions = sgd_clf.predict(test)
### START CODING HERE ###

In [39]:
type(test_predictions)

numpy.ndarray

In [40]:
print(test_predictions.shape)

(171,)


Answer the following questions HERE: 

Q1 - What is the data type of test_pred? Add a new cell above this and type your code, and then type your answer here.

test_pred is type numpy.ndarray

Q2 - What is the shape of your test_pred? Why? Add another cell above this and type your code to get the answer and then explain your answer here.

The shape is one-dimensional with 171 elements, because it is just the model's predictions about each individual data point in test.

## <font color="red"> Required Coding

In [43]:
### START CODING HERE ###

# Import the necessary module from sklearn for accuracy_score and 
from sklearn.model_selection import cross_val_score

# compute the accuracy, store it in acc variable
acc = cross_val_score(sgd_clf, train, train_labels, cv=3, scoring="recall")

# Print the accuracy 
print(acc)

### END CODING HERE ###

[0.86746988 0.31325301 0.48192771]


Answer the following questions HERE:

Q3 - What is your accuracy in %? 

Above 77%

Q4 - Change your accuracy to precision and type the result here.

Above 90%

Q5 - Change the precision to recall and type the results here.

Above 65%

Q6 - Compare precision and recall. Which one do you think is more important in cancer diagnosis?

Precision is how well the model identifies a positive and does not give false positives. Recall is how well the model identifies positives and does not miss any. Recall is more important in cancer diagnosis because the consequences of missing something are far worse (could be death) than the consequences of incorrectly labeling something as cancerous (likely just stressful to the patient until obtaining a second opinion.  

Q7 - Change the random_state of your classifier and try three different values. Compare and discuss your results. Can you improve the accuracy by changing that parameter? Can you improve the recall without affecting precision? Explain! 

random_state = 420
accuracy above 84%
precision above 80%
recall above 80%

random_state = 4200
accuracy above 83%
precision above 92%
recall above 77%

random_state = 42000
accuracy above 57%
precision above 91%
recall above 31%

Changing the random_state can affect accuracy, precision, and recall as the above shows. You cannot improve either recall or precision without negatively affecting the other measure. This is because your model will favor one type of error over another (either false positives or false negatives) which precision and recall are measures of. They are inversely related.

## Part II - Regression and Overfitting

Now, we want to display how overfitting occurs in regression. First, run the following cell.

In [44]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score


def true_fun(X):
    return np.cos(1.5 * np.pi * X)

np.random.seed(0)

n_samples = 30
degrees = [1, 4, 15]

X = np.sort(np.random.rand(n_samples))
y = true_fun(X) + np.random.randn(n_samples) * 0.1

plt.figure(figsize=(14, 5))
for i in range(len(degrees)):
    ax = plt.subplot(1, len(degrees), i + 1)
    plt.setp(ax, xticks=(), yticks=())

    polynomial_features = PolynomialFeatures(degree=degrees[i],
                                             include_bias=False)
    linear_regression = LinearRegression()
    pipeline = Pipeline([("polynomial_features", polynomial_features),
                         ("linear_regression", linear_regression)])
    pipeline.fit(X[:, np.newaxis], y)

    # Evaluate the models using crossvalidation
    scores = cross_val_score(pipeline, X[:, np.newaxis], y,
                             scoring="neg_mean_squared_error", cv=10)

    X_test = np.linspace(0, 1, 100)
    plt.plot(X_test, pipeline.predict(X_test[:, np.newaxis]), label="Model")
    plt.plot(X_test, true_fun(X_test), label="True function")
    plt.scatter(X, y, edgecolor='b', s=20, label="Samples")
    plt.xlabel("x")
    plt.ylabel("y")
    plt.xlim((0, 1))
    plt.ylim((-2, 2))
    plt.legend(loc="best")
    plt.title("Degree {}\nMSE = {:.2e}(+/- {:.2e})".format(
        degrees[i], -scores.mean(), scores.std()))
plt.show()


<Figure size 1400x500 with 3 Axes>

Answer the following questions HERE:

Q1 - How do you compare the three plots above? 

Mean Squared Area for the plots can be compared to see that the center plot is the closest to the data. 

Q2 - Which one has overfitting? Why?

The last plot has overfitting, because you can see that the samples lie on or near the model, but you can also see that the model in it's entirety is dramatically far from the true function. 

## Grading

For assignment 1, your notebook will be run and graded manually with a maximum of 100 points. Make sure that you get the desired outputs for all cells that you implemented. Also, your notebook should be written with no grammatical and spelling errors and should be nicely-formatted and easy-to-read.

The breakdown of the 100 points is as follows:

Part I has 80 points:
- 45 points: correct implementation
- 35 points: correct answers to 7 questions

Part II has 10 points:

The remaining 10 points will be based on your writing and formatting as instructed in the notebook.  Follow the instructions of each section carefully. Points will be deducted if your submitted notebook is not easy to read and follow or if it has grammatical and spelling errors.